x/exchange/keeper/fulfillment.go

```diff
@@ -0,0 +1,375 @@
+func validateAcceptingOrdersAndCanUserSettle(store sdk.KVStore, marketID uint32) error {
+	if err := validateMarketIsAcceptingOrders(store, marketID); err != nil {
+		return err
+	}
+	if !isUserSettlementAllowed(store, marketID) {
+		return fmt.Errorf("market %d does not allow user settlement", marketID)
+	}
+	return nil
+}
+
+// FillBids settles one or more bid orders for a seller.
+func (k Keeper) FillBids(ctx sdk.Context, msg *exchange.MsgFillBidsRequest) error {
+	if err := msg.ValidateBasic(); err != nil {
+		return err
+	}
+
+	marketID := msg.MarketId
+	store := k.getStore(ctx)
+
+	if err := validateAcceptingOrdersAndCanUserSettle(store, marketID); err != nil {
+		return err
+	}
+	seller := sdk.MustAccAddressFromBech32(msg.Seller)
+	if err := k.validateUserCanCreateAsk(ctx, marketID, seller); err != nil {
+		return err
+	}
+	if err := validateCreateAskFees(store, marketID, msg.AskOrderCreationFee, msg.SellerSettlementFlatFee); err != nil {
+		return err
+	}
+
+	orders, oerrs := k.getBidOrders(store, marketID, msg.BidOrderIds, msg.Seller)
+	if oerrs != nil {
+		return oerrs
+	}
+
+	var errs []error
+	var totalSellerFee sdk.Coins
+	assetOutputs := make([]banktypes.Output, 0, len(msg.BidOrderIds))
+	priceInputs := make([]banktypes.Input, 0, len(msg.BidOrderIds))
+	addrIndex := make(map[string]int, len(msg.BidOrderIds))
+	feeInputs := make([]banktypes.Input, 0, len(msg.BidOrderIds)+1)
+	feeAddrIndex := make(map[string]int, len(msg.BidOrderIds))
+	filledOrders := make([]*exchange.FilledOrder, 0, len(msg.BidOrderIds))
+	for _, order := range orders {
+		bidOrder := order.GetBidOrder()
+		buyer := bidOrder.Buyer
+		assets := bidOrder.Assets
+		price := bidOrder.Price
+		buyerSettlementFees := bidOrder.BuyerSettlementFees
+
+		sellerRatioFee, rerr := calculateSellerSettlementRatioFee(store, marketID, price)
+		if rerr != nil {
+			errs = append(errs, fmt.Errorf("error calculating seller settlement ratio fee for order %d: %w",
+				order.OrderId, rerr))
+			continue
+		}
+		if err := k.releaseHoldOnOrder(ctx, order); err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		if sellerRatioFee != nil {
+			totalSellerFee = totalSellerFee.Add(*sellerRatioFee)
+		}
+
+		i, seen := addrIndex[buyer]
+		if !seen {
+			i = len(assetOutputs)
+			addrIndex[buyer] = i
+			assetOutputs = append(assetOutputs, banktypes.Output{Address: buyer})
+			priceInputs = append(priceInputs, banktypes.Input{Address: buyer})
+		}
+		assetOutputs[i].Coins = assetOutputs[i].Coins.Add(assets)
+		priceInputs[i].Coins = priceInputs[i].Coins.Add(price)
+
+		if !buyerSettlementFees.IsZero() {
+			j, known := feeAddrIndex[buyer]
+			if !known {
+				j = len(feeInputs)
+				feeAddrIndex[buyer] = j
+				feeInputs = append(feeInputs, banktypes.Input{Address: buyer})
+			}
+			feeInputs[j].Coins = feeInputs[j].Coins.Add(buyerSettlementFees...)
+		}
+
+		filledOrders = append(filledOrders, exchange.NewFilledOrder(order, price, buyerSettlementFees))
+	}
```

Comment:
This whole block seems pretty self contained and you can probably extract it out to make FillBids shorter and easier to understand.